### PR TITLE
mat: add minimal shadow detection for complex matrices

### PR DESCRIPTION
**Pull request by kortschak**
This is initial bootstrapping work to allow generation of CDense from Dense.

I'm not sure of `RawCMatrix`. Is this what we want?

Please take a look.

### CODE REVIEW

File: mat/offset.go [hunk 1]
```diff
@@ -18,3 +18,14 @@ func offset(a, b []float64) int {
...
+
+// offsetComplex returns the number of complex128 values b[0] is after a[0].
+func offsetComplex(a, b []complex128) int {
+	if &a[0] == &b[0] {
+		return 0
+	}
+	// This expression must be atomic with respect to GC moves.
+	// At this stage this is true, because the GC does not
+	// move. See https://golang.org/issue/12445.
+	return int(uintptr(unsafe.Pointer(&b[0]))-uintptr(unsafe.Pointer(&a[0]))) / int(unsafe.Sizeof(float64(0)))
```
Comment (vladimir-ch) [inline mat/offset.go]: Shouldn't it be `complex128`?

File: mat/offset_appengine.go [hunk 2]
```diff
@@ -22,3 +22,16 @@ func offset(a, b []float64) int {
...
+func offsetComplex(a, b []complex128) int {
+	va0 := reflect.ValueOf(a).Index(0)
+	vb0 := reflect.ValueOf(b).Index(0)
+	if va0.Addr() == vb0.Addr() {
+		return 0
+	}
+	// This expression must be atomic with respect to GC moves.
+	// At this stage this is true, because the GC does not
+	// move. See https://golang.org/issue/12445.
+	return int(vb0.UnsafeAddr()-va0.UnsafeAddr()) / sizeOfFloat64
```
Comment (vladimir-ch) [inline mat/offset_appengine.go]: Also here?

File: mat/offset.go (hunk 1 above)
Comment (kortschak) [reply mat/offset.go]: Yes. Thanks

File: mat/offset_appengine.go (hunk 2 above)
Comment (kortschak) [reply mat/offset_appengine.go]: Fixed.

**Comment by btracey** [comment]
I haven't fully finished reviewing the shadow_complex bit but that's all I noticed as well.

**Comment by kortschak** [comment]
Yeah. It's essentially copy/paste, with all the errors associated with that approach. There is no testing because that depends on `Slice`. It will be very much easier when there is a reasonable set of generated code for complex.

File: mat/offset.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (btracey) [reply mat/offset.go]: Whoops, didn't realize these were still hanging around

**Pull request merged**